traversal/GraphTraversal.java

```diff
@@ -130,154 +131,108 @@ public Type() {
         }
     }
 
     public static class Thing extends GraphTraversal {
 
         private Planner planner;
         private TraversalCache cache;
-        private boolean modifiable;
 
         public Thing() {
             super();
-            modifiable = true;
         }
 
         public void initialise(TraversalCache cache) {
             assert planner == null;
             this.cache = cache;
-            planner = this.cache.getPlanner(structure, Planner::create);
+            planner = this.cache.getPlanner(structure, modifiers, sm -> Planner.create(sm.first(), sm.second()));
         }
 
         @Override
         FunctionalIterator<VertexMap> permutationIterator(GraphManager graphMgr) {
             assert planner != null && cache != null;
-            FunctionalIterator<VertexMap> iter = permutationIterator(graphMgr, planner, false, filter());
-            cache.mayUpdatePlanner(structure, planner);
+            FunctionalIterator<VertexMap> iter = permutationIterator(graphMgr, planner, false);
+            cache.mayUpdatePlanner(structure, modifiers, planner);
             return iter;
         }
 
         FunctionalProducer<VertexMap> permutationProducer(GraphManager graphMgr,
                                                           int parallelisation) {
             assert planner != null && cache != null;
             planner.tryOptimise(graphMgr, false);
-            FunctionalProducer<VertexMap> producer = planner.procedure().producer(graphMgr, parameters, filter(), parallelisation);
-            cache.mayUpdatePlanner(structure, planner);
+            FunctionalProducer<VertexMap> producer = planner.procedure().producer(graphMgr, parameters, modifiers, parallelisation);
+            cache.mayUpdatePlanner(structure, modifiers, planner);
             return producer;
         }
 
-        // TODO: We should not dynamically calculate properties like this, and then guard against 'modifiable'.
-        //       We should introduce a "builder pattern" to Traversal, such that users of this library will build
-        //       traversals with Traversal.Builder, and call .build() in the end to produce a final Object.
-        @Override
-        Set<Identifier.Variable.Retrievable> filter() {
-            if (filter.isEmpty()) {
-                modifiable = false;
-                iterate(structure.vertices()).filter(v -> v.id().isRetrievable()).map(v -> v.id().asVariable().asRetrievable())
-                        .toSet(filter);
-            }
-            return filter;
-        }
-
-        @Override
-        public void filter(Set<? extends Identifier.Variable.Retrievable> filter) {
-            assert modifiable && iterate(filter).noneMatch(Identifier::isLabel);
-            this.filter.addAll(filter);
-        }
-
         public void equalThings(Identifier.Variable thing1, Identifier.Variable thing2) {
-            assert modifiable;
```

Comment:
using this to protect us against using the class as a Builder object was incorrect anyway, and now the place where we set `modifiable` to false is no longer existing. Instead, we delete the check and move the TODO to the top level of the class, where it should be implemented for all types of `GraphTraversal`